### PR TITLE
Add new functionality to binarize preference datasets directly from distilabel

### LINKED ISSUE #263: [FEATURE] Add functionality to binarize datasets for fine-tuning directly from distilabel
**Is your feature request related to a problem? Please describe.**
We need helper functions to transform our preference datasets to the format expected for trainers like [DPOTrainer](https://huggingface.co/docs/trl/main/en/dpo_trainer).

**Describe the solution you'd like**
Functions to transform the dataset from a `Pipeline` in a format suitable for fine-tuning alignment.

**Describe alternatives you've considered**
Do it by hand.

**Additional context**

### PR DESCRIPTION
**Pull request by plaguss**
## Description

This PR adds a function to binarize a `CustomDataset` with a `PreferenceTask`:

```python
from distilabel.utils import prepare_dataset
from datasets import load_dataset
from distilabel.tasks import JudgeLMTask
from distilabel.dataset import prepare_dataset

dataset = load_dataset("argilla/distilabel-intel-orca-dpo-pairs", split="train")
# Add the task that originally was used for the labelling
dataset.task = JudgeLMTask()
dataset_binarized_random = prepare_dataset(dataset, strategy="random", keep_ties=True)
```

Closes #263

### CODE REVIEW

File: src/distilabel/utils/dataset.py [hunk 1]
```diff
@@ -51,3 +60,215 @@ def load_task_from_disk(path: Path) -> "Task":
     with open(task_path, "rb") as f:
         task = pickle.loads(f.read())
     return task
+
+
+def _binarize_dataset(
+    dataset: "CustomDataset",
+    seed: int = None,
+    strategy: BinarizationStrategies = "random",
+    keep_ties: bool = True,
```
Comment (davidberenstein1957) [inline src/distilabel/utils/dataset.py:69]: why did we set `keep_ties` to `True` as default? Intuitively I would say removing them will lead to higher quality data.

File: src/distilabel/utils/dataset.py [hunk 2]
```diff
@@ -51,3 +60,215 @@ def load_task_from_disk(path: Path) -> "Task":
...
+    **kwargs: Any,
+) -> "CustomDataset":
+    """Binarizes a distilabel dataset.
```
Comment (davidberenstein1957) [inline src/distilabel/utils/dataset.py:72]: again, some people might be like: "what is binarizing?"

File: src/distilabel/utils/dataset.py [hunk 3]
```diff
@@ -51,3 +60,215 @@ def load_task_from_disk(path: Path) -> "Task":
...
+    Raises:
+        ValueError: If the strategy is not implemented.
+
+    Returns:
+        CustomDataset: Dataset binarized.
+    """
+    rating_column = "rating"
+    responses_column = "generations"
+
+    def binarize_random(example):
```
Comment (davidberenstein1957) [inline src/distilabel/utils/dataset.py:92]: there seems to be some code duplication in binarize_random and binarize_worst, perhaps we can combine these a bit more?

Comment (davidberenstein1957) [inline src/distilabel/utils/dataset.py:92]: type hints

File: src/distilabel/utils/dataset.py [hunk 4]
```diff
@@ -51,3 +60,215 @@ def load_task_from_disk(path: Path) -> "Task":
...
+            "chosen": chosen_response,
+            "rejected": random_response,
+            "rating_chosen": int(best_rating),
+            "rating_rejected": int(random_rating),
+            "chosen_model": chosen_model,
+            "rejected_model": random_model,
+        }
+        return binarized
+
+    def binarize_worst(example):
```
Comment (davidberenstein1957) [inline src/distilabel/utils/dataset.py:137]: time hints?

File: src/distilabel/utils/dataset.py [hunk 5]
```diff
@@ -51,3 +60,215 @@ def load_task_from_disk(path: Path) -> "Task":
...
+            lambda example: example["rating_chosen"] != example["rating_rejected"]
+        )
+    return dataset
+
+
+def prepare_dataset(
+    dataset: "CustomDataset",
+    strategy: BinarizationStrategies = "random",
+    seed: Optional[int] = None,
+    keep_ties: bool = True,
```
Comment (davidberenstein1957) [inline src/distilabel/utils/dataset.py:193]: same here w.r.t. keep_ties=True

File: src/distilabel/utils/dataset.py (hunk 1 above)
Comment (plaguss) [reply src/distilabel/utils/dataset.py:69]: Wasn't sure what value to set as default, and thought that allowed to keep all the data initially, but yes, I think it makes more sense to remove those by default

File: src/distilabel/utils/dataset.py (hunk 3 above)
Comment (plaguss) [reply src/distilabel/utils/dataset.py:92]: sure will take a look

Comment (plaguss) [reply src/distilabel/utils/dataset.py:92]: they are not that usual in internal functions, and that's the standard example from [HuggingFace's datasets](https://huggingface.co/docs/datasets/process#map).

File: src/distilabel/utils/dataset.py (hunk 4 above)
Comment (plaguss) [reply src/distilabel/utils/dataset.py:137]: same as `binarize_random`

File: src/distilabel/utils/dataset.py (hunk 2 above)
Comment (plaguss) [reply src/distilabel/utils/dataset.py:107]: Could you both @davidberenstein1957 and @dvsrepo take a look at the new section of the docs for the dataset preparation/binarization? This commit https://github.com/argilla-io/distilabel/pull/264/commits/e4602d4e2c0a68737915cfe6a743ead343091d4e contains the updates.

**Comment by dvsrepo** [comment]
@plaguss it looks good, I don't have cycles to do an in-depth review.

One suggestion: would it be possible to return the chosen and rejected in the OpenAI format, like this?

https://huggingface.co/datasets/HuggingFaceH4/ultrafeedback_binarized?row=0

I think this would make it more interoperable

**Comment by plaguss** [comment]
> @plaguss it looks good, I don't have cycles to do an in-depth review.
> 
> One suggestion: would it be possible to return the chosen and rejected in the OpenAI format, like this?
> 
> https://huggingface.co/datasets/HuggingFaceH4/ultrafeedback_binarized?row=0
> 
> I think this would make it more interoperable

It should be easy, will take a look and update the code

**Review by davidberenstein1957 — APPROVED** [review]

**Comment by davidberenstein1957** [comment]
@plaguss can you resolve the merge conflicts. After that you can merge it. @sdiazlor , this might be interesting to include or mention in your tutorial too. https://github.com/argilla-io/distilabel/issues/247

**Pull request merged**